### ISSUE TITLE
Max Reduce With Indices Generality

### ISSUE THREAD

**Issue by wransom-TT** [tenstorrent/tt-metal#28141]
Recently the `max_reduce_with_indices` was implemented as a more performant version of TopK specific to the Max Pool With Indices use case.  Currently it only supports kernel sizes equal to 9, Uint16 indexes, and 1 tile per reduction.  We will need the following support:
- Kernel sizes <= 32 (for generality)
- Uint32 index support (for generality)
- Mutli-tile support (for performance)

**Comment by vmilicevicTT** [comment]
potentially assign to someone from ops team @cmaryanTT

**Comment by rdjogoTT** [comment]
Implementation note: (*Came to mind after the MPWI presentation I gave*)
- Rather than loading 8 rows of odd and even columns into LREGs 0-3, we can load 16 rows of just even columns (the just off next iteration)
- This will reduce the number of intermediates that need to be stored back to the Dest reg which will reduce number of instrs. overall

Pseudocode (for even cols):
```
load(lreg = 0, address = 0)
load(lreg = 1, address = 4)
load(lreg = 2, address = 8)
load(lreg = 3, address = 12)
transpose()
swap(lreg0, lreg1)
swap(lreg2, lreg3)
swap(lreg0, lreg2)
transpose()
swap(lreg0, lreg1)
swap(lreg2, lreg3)
swap(lreg0, lreg2)
store(lreg = 0, address = 0)
```
-> this will sort the first 16 rows of even cols (whole F0)

**Comment by fahmedchanTT** [comment]
@wransom-TT Can you please provide an update on this issue's progress

**Comment by wransom-TT** [comment]
@fahmedchanTT work has not started on this issue

**Comment by skhatriTT** [comment]
@wransom-TT could you please provide an ETA for this issue? This will help us align, as it’s currently blocking all CNN model.

**Comment by wransom-TT** [comment]
@skhatriTT can you provide me more insight on what max pool traces are blocked currently? There are several different generality limitations currently, some easier than others to fix.

**Comment by abogdanovicTT** [comment]
While running tests for CNN models, we discovered that the following models are currently failing due to the max_pool_with_indices op (more models will be affected once we run all Forge models in training mode in CI):

```
pytorch_wide_resnet_wide_resnet50_2_cv_image_cls_torch_hub_0_4_shlo_after_stablehlo.mlir
pytorch_vgg_torchvision_vgg11_bn_cv_image_cls_torchvision_0_2_shlo.mlir
pytorch_rcnn_alexnet_cv_object_det_torchvision_0_3_shlo_after_frontend_stablehlo.mlir
pytorch_wide_resnet_wide_resnet50_2_cv_image_cls_torch_hub_0_3_shlo_after_frontend_stablehlo.mlir
pytorch_resnet_resnet101_cv_image_cls_torchvision_0_3_shlo_after_frontend_stablehlo.mlir
pytorch_resnet_resnet101_cv_image_cls_torchvision_0_4_shlo_after_stablehlo.mlir
pytorch_vgg_torchvision_vgg11_bn_cv_image_cls_torchvision_0_4_shlo_after_stablehlo.mlir
pytorch_vgg_torchvision_vgg11_bn_cv_image_cls_torchvision_0_3_shlo_after_frontend_stablehlo.mlir
pytorch_resnet_resnet101_cv_image_cls_torchvision_0_2_shlo.mlir
pytorch_rcnn_alexnet_cv_object_det_torchvision_0_4_shlo_after_stablehlo.mlir
pytorch_wide_resnet_wide_resnet50_2_cv_image_cls_torch_hub_0_2_shlo.mlir
pytorch_rcnn_alexnet_cv_object_det_torchvision_0_2_shlo.mlir
```

In [max_pool2d_metal_shapes.csv](https://github.com/user-attachments/files/22743800/max_pool2d_metal_shapes.csv) are the shapes and other arguments used in the ttnn.max_pool2d function, for the listed models. 
Based on this test example, the output values, indices, and shapes are correct only for a kernel size of [3, 3]; other kernel sizes are not supported.


```
import ttnn
import torch

device = ttnn.open_device(device_id=0, l1_small_size=2800)
input_tensor = torch.randn(1, 64, 112, 112, dtype=torch.bfloat16)
ttnn_input_shape = (1, 1, 112 * 112, 64)
torch_input_permuted = torch.permute(input_tensor, (0, 2, 3, 1)).reshape(ttnn_input_shape)  
tt_input = ttnn.from_torch(torch_input_permuted, ttnn.bfloat16, layout=ttnn.ROW_MAJOR_LAYOUT, device=device)

output, indices = ttnn.max_pool2d(
    tt_input,
    batch_size=1,
    input_h=112,
    input_w=112,
    channels=64,
    kernel_size=[3, 3],    
    stride=[2, 2],
    padding=[1, 1, 1, 1],
    dilation=[1, 1],
    ceil_mode=False,
    return_indices=True
)

torch_output = ttnn.to_torch(output, device=device)
torch_indices = ttnn.to_torch(indices, device=device)
print("Output shape:", torch_output.reshape(1, 56, 56, 64).permute(0, 3, 1, 2).shape)    
print("Indices shape:", torch_indices.reshape(1, 56, 56, 64).permute(0, 3, 1, 2).shape) 
```

**Comment by wransom-TT** [comment]
@abogdanovicTT @skhatriTT ok, so the major MPWI limitations currently are:
1. Tensor HxW must be less than 65536 (256x256)
2. Kernels must be equal to 9 sticks
3. Kernels must be < 32 sticks (so for square kernels 5x5 or smaller)

Numbers 2 and 3 are related, but actually different limitations caused by very different issues.

Number 2 also blocks number 3.

So we can either start working on Number 1 or number 2.  Which of these is higher priority for your models?  It sounds like its the kernel size limitation (number 2)?

**Comment by vladimirjovanovicTT** [comment]
It seems that number 2 (kernel shapes) is what we need, based on our sheet. @wransom-TT

**Comment by wransom-TT** [comment]
@vladimirjovanovicTT ok sounds good. for that one we need a new LLK function implementation. The LLK team is pretty backed up but we have discussed offloading this effort to the Conv team. We will get started on this asap and I will let you know when we have an ETA.

**Comment by wransom-TT** [comment]
@vladimirjovanovicTT also depending on how long the general kernel size LLK implementation will take it should also be possible to hack the kernel=9 case to work for kernel=4 with some padding so I'll weigh that option once we get a pulse on the LOE for the general implementation.

**Comment by vladimirjovanovicTT** [comment]
@wransom-TT thanks; regarding the hack to support kernel=4, please let us know about the ETAs for different options so we can coordinate with the effort on compiler side.

**Comment by wransom-TT** [comment]
@vladimirjovanovicTT the hack was very straightforward, so I went ahead and implemented this here:

https://github.com/tenstorrent/tt-metal/pull/30329

Please let me know if this unblocks you.  NOTE: this should work for arbitrary C dimensions on WH, but currently BH only supports C <= 16, which will be fixed in an open PR hopefully getting merged tomorrow or early next week.

We will be discussing a proper solution for general kernel sizes with the LLK team this Friday with work beginning early next week at which point I can provide an estimated ETA.

**Comment by vladimirjovanovicTT** [comment]
Hey @wransom-TT, we tested this and it works well for [2,2] kernels.
I have heard that there is an effort to support larger kernels (up to tile size); is there a plan/roadmap for this functionality?

**Comment by wransom-TT** [comment]
@vladimirjovanovicTT that's good to hear! We are beginning work on the LLK function necessary to process kernels up to 32 sticks today, but I don't have an ETA for this just yet.

**Comment by wransom-TT** [comment]
This is a duplicate of https://github.com/tenstorrent/tt-metal/issues/27845
